### ISSUE TITLE
Throw an exception when Record.field(Field) runs into ambiguous columns

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#4477]
As a next step after #4476

**Comment by kaiwaldron** [comment]
I see this is targeted for for 4.0.0 because it is a breaking change. In the meantime, is there an alternative way to get a field from a Record by an exact match (of both table and field name)? The fallback logic to just matching by field name is causing issues.

**Comment by lukaseder** [comment]
See, this "exact match" is a bit treacherous. For instance, why do you think the table name is relevant but not the schema name, nor the catalog name? What's the use-case for this specific behaviour?

Note that this issue isn't about exact matching. Ambiguous columns are defined by the fact that two columns are really *the same* (i.e. they both have exactly the same field name, or table name *and* field name, etc.). For example:

```sql
SELECT id, id
FROM book
```

In any case, I would prefer not to add anything like that right now. However, you could work around the issue by using `Record.intoMap()` or something like that.

**Comment by RichardBradley** [comment]
I have been bitten by this issue. I have a query which joins a few tables and some separate view code which fetches the data from the record:

```
Record r = dsl
   .select(
      PARENT.ID,
      PARENT.NAME,
      CHILD.NAME,
      CHILD2.SIZE)
    .from(PARENT)
    .join(CHILD).onKey()
    .join(CHILD2).onKey()
```

If, due to a programming error, I make a get which is "ambiguous", I see only a silent warning in the logs, when in many cases incorrect data is returned:

```CHILD2.NAME.get(record) // this returns CHILD.NAME from the record, with a warning in the logs```

This issue would turn that into an exception rather than a warning, which I support.

Further to that, if, due to a programming error, I make a get which is incorrect but not seen as "ambiguous" by `Fields.field(Field<T> field)`, then JOOQ will return incorrect data with no warning:

```CHILD.ID.get(record) // this returns PARENT.ID from the record without warning```

**Comment by lukaseder** [comment]
Thank you for your feedback, @RichardBradley. 

> This issue would turn that into an exception rather than a warning, which I support.

Yes, this will definitely be improved.

> I make a get which is incorrect but not seen as "ambiguous" by Fields.field(Field<T> field), then JOOQ will return incorrect data with no warning:

This, on the other hand, works as designed, and while I understand that in your case, the behaviour looks incorrect, there's a very convenient feature in there that produces this behaviour, namely the fact that you can alias a table and still access the columns through the original table reference, if so desired:

```java
Table<?> a2 = A1.as("a2");
Field<?> a2Id = a2.field(A1.ID);
```

This behaviour makes a lot of sense in a variety of use-cases. The price to pay is the fact that there might be an accidental match, indeed.

Hope this helps

**Comment by dweitzman** [comment]
Hello! I'd like be be able to propose using jooq for a project with lots of developers of varying levels of experience, but the lack of type safety in this scenario makes me nervous. I can easily imagine someone having misplaced confidence in the type safety of jooq and getting burned by this, similar to Richard's experience mentioned above.

Maybe there's a different way to handle aliases where the field selectors can become aware of a table alias? Or if not, maybe it's worth making aliasing a less convenient to maintain type safety and protect against (inevitable) programming errors?

**Comment by RichardBradley** [comment]
I have written some code which works around this issue for now, which you can find at https://gist.github.com/RichardBradley/676ef35db100532bbc32a3b2344a1723 . I hope this is helpful for others with the same problem.

This code (which is specific to `log4j2` but should be easily translatable to other loggers) does two things: 
 1) listens for the "ambiguous column" warning log message and throws, as I consider this an error rather than a warning
 2) wraps the "Record" objects in a proxy which detects some ambiguous column matches which are not currently warned on and throws an exception when the above bug might be occurring. Of course, you have to remember to wrap all Record instances in this helper to get the benefit, so a proper fix here would be better for me.

Hope this helps!

**Comment by lukaseder** [comment]
> I can easily imagine someone having misplaced confidence in the type safety of jooq and getting burned by this, similar to Richard's experience mentioned above.

There are two things to consider:

1. Yes, this was an early design mistake, and we'll fix it eventually, just not in a minor release
2. Type safety in an internal DSL is always a tradeoff. For instance, you can easily form a predicate that references a column from a table that is not part of the `FROM` clause.

So, I recommend not misplacing confidence yet. jOOQ works very well in most cases, also for junior developers.

> Maybe there's a different way to handle aliases where the field selectors can become aware of a table alias? Or if not, maybe it's worth making aliasing a less convenient to maintain type safety and protect against (inevitable) programming errors?

Would you mind describing these ideas? Keep in mind that any tradeoff in this area might drastically reduce the convenience of jOOQ API usage, and is of course also backwards-incompatible...

**Comment by dweitzman** [comment]
What about a system property to disallow mismatched table aliases for those
who prefer that behavior?

For folks using that setting, maybe a Field.inTable() method that could
adapt a field to a different table or remove the table name if you want
ambiguous matching?

For our use case we wouldn't be too worried about the alias situation as
long as we had the system property

**Comment by lukaseder** [comment]
System properties are hard to maintain and document. I mean, their global, static nature of course has the benefit of being applicable *everywhere*, but imagine two instances of jOOQ running on the same JVM where one relies on this property being set?

We have `Settings` inside of `Configuration` for these things, but then again, `Settings` rely on some instant context, which cannot always be provided to a `Table`.

I guess there's no proper solution short of simply breaking the behaviour incompatibly. But before we rush things, I'd like to better understand your worries. Sure, this is a sore spot of the API, but then again, there are many other kinds of mistakes that can happen, for instance when you transfer your records to maps. I'd like to really understand your sense of urgency here first, prior to do any changes. May you please explain?

**Comment by RichardBradley** [comment]
> I'd like to really understand your sense of urgency here first, prior to do any changes. May you please explain?

This caused an actual bug in production for my team. See [my notes above](https://github.com/jOOQ/jOOQ/issues/4477#issuecomment-308483636). 

Perhaps we are not using jOOQ in the way you envisaged. Currently, we have some "DAO" classes which take domain classes like "filters" and business rules and turn them into somewhat complex SQL queries using jOOQ. The DAOs then return Records which represent mutliple tables joined together. Separately, we have "view" type classes which convert those Records into domain objects for use in the rest of our app. These mapping classes were fetching column data using the jOOQ `Field` classes but jOOQ was returning incorrect data due to the table name mismatch discussed here (with a WARN log that no-one noticed).

We don't seem to be using the table alias functionality that conflicts with this change much or at all; we certainly haven't missed it since I effectively disabled it with the JooqContext code [linked above](https://github.com/jOOQ/jOOQ/issues/4477#issuecomment-340191976).

I am not able to share our application code here, sorry, but I can answer any questions you have about our usage.

> > Maybe there's a different way to handle aliases where the field selectors can become aware of a table alias?
> Would you mind describing these ideas? 

I think something along these lines might be able to satisfy both camps.

Perhaps something like:

```
Table<Foo> fooAsBar = Tables.FOO.as("bar");
Field<String> barName = fooAsBar.NAME; // currently this field has table name "foo", 
                                       // but we'd like it to have table name "bar" somehow

// maybe:
Field<String> barName = fooAsBar.nameField(); // a method that passes along the aliased name?
Field<String> barName = fooAsBar.name; // a non-static field, rather than a static NAME field, to
                                       // pass along the aliased name?

```

**Comment by dweitzman** [comment]
The thing I'm afraid of would look pretty much the same as Richard's example. In our db there are many fields names which appear in multiple tables. The most common one would be "id", similar to Richard, although there are others that are equally important not to mix up. The approximate test I ran:

```
void businessLogic() {
  Record joinedRecord = fetchRecord();
  BarRecord barRecord = joinedRecord.into(BarRecord.class); 
  // Should raise an exception, or else there's a risk of data corruption or failing to
  // run business logic correctly.
  long NOT_REALLY_THE_BAR_ID = barRecord.getId();
}

Record fetchRecord() {
  return select(FOO.ID, BAR.NAME).from(FOO.join(BAR).on(FOO.ID.eq(BAR.ID))).fetchOne();
}
```

This code seems easy to write by accident, easy to miss in code review, and the consequences could be severe: data corruption and/or failure to run essential business logic.

**Comment by McNetic** [comment]
When making the move, it should be considered that (as far as I know) the only way to check for the existance of a specific field in a record is to compare it to itself like this:
``` TABLE.FIELD.equals(record.field(TABLE.FIELD)) ```
While it is possible to catch the resulting exception, i would be much nicer of there were a dedicated API for checking for the existance of a field in a record.

**Comment by lukaseder** [comment]
@McNetic Thanks for commenting. I'm assuming that you're looking for a check that returns true only if the field has the exact same name *and* qualification. What's a use-case for such a check?

**Comment by McNetic** [comment]
The use case is to check if the field exists in a record (see https://groups.google.com/forum/#!topic/jooq-user/5ziIzqgWYQw).

**Comment by jimmyzzxhlh** [comment]
> The use case is to check if the field exists in a record (see https://groups.google.com/forum/#!topic/jooq-user/5ziIzqgWYQw).

+1 on this. I created a helper function like this which checks if the field exists, and I thought this should work when joining multiple tables, but it doesn't. 

```
  public static <T> void setIfPresent(Record record, Field<T> field, Consumer<T> setter) {
    if (record.field(field) != null) {
      Optional.ofNullable(record.get(field))
          .ifPresent(setter);
    }
  }
```

which then I can use it to convert Jooq Record into Protobufs:

```
  private static PrincipalAgent buildPrincipalAgentFromRecord(Record record) {
    PrincipalAgent.Builder builder = PrincipalAgent.newBuilder();

    setIfPresent(record, AGENT.ID, value -> builder.setAgentId(buildInt64Value(value)));
    setIfPresent(record, AGENT.FIRST_NAME, value -> builder.setFirstName(buildStringValue(value)));
    setIfPresent(record, AGENT.MIDDLE_NAME, value -> builder.setMiddleName(buildStringValue(value)));
    setIfPresent(record, AGENT.LAST_NAME, value -> builder.setLastName(buildStringValue(value)));

    return builder.build();
  }
```

But, I was bitten by the exact same issue mentioned above, because my SQL query was selecting a different ID other than the `AGENT.ID`. Therefore, I'm setting a different ID in the `setIfPresent` function without getting any exceptions. This bug can be very subtle and hard to detect, if there are overlapping IDs across different tables. 

IMHO this can be solved by providing a function that returns an exact match `Field` object.
